src/frontend/src/optimizer/plan_node/stream_share.rs

```diff
@@ -0,0 +1,137 @@
+use risingwave_pb::stream_plan::{
+    DispatchStrategy, DispatcherType, ExchangeNode, StreamNode as ProstStreamPlan,
+};
+
+use super::{PlanRef, PlanTreeNodeUnary, StreamNode};
+use crate::optimizer::plan_node::{LogicalShare, PlanBase, PlanTreeNode};
+use crate::optimizer::property::Distribution;
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+#[derive(Debug, Clone)]
+pub struct StreamShare {
+    pub base: PlanBase,
+    logical: LogicalShare,
+}
+
+impl StreamShare {
+    pub fn new(logical: LogicalShare) -> Self {
+        let ctx = logical.base.ctx.clone();
+        let input = logical.input();
+        let pk_indices = logical.base.logical_pk.to_vec();
+        let dist = input.distribution().clone();
+        // Filter executor won't change the append-only behavior of the stream.
+        let base = PlanBase::new_stream(
+            ctx,
+            logical.schema().clone(),
+            pk_indices,
+            logical.functional_dependency().clone(),
+            dist,
+            logical.input().append_only(),
+        );
+        StreamShare { base, logical }
+    }
+}
+
+impl fmt::Display for StreamShare {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.logical.fmt_with_name(f, "StreamShare")
+    }
+}
+
+impl PlanTreeNodeUnary for StreamShare {
+    fn input(&self) -> PlanRef {
+        self.logical.input()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::new(self.logical.clone_with_input(input))
+    }
+}
+
+impl_plan_tree_node_for_unary! { StreamShare }
+
+impl StreamNode for StreamShare {
+    fn to_stream_prost_body(&self, _state: &mut BuildFragmentGraphState) -> ProstStreamNode {
+        unreachable!("stream scan cannot be converted into a prost body -- call `adhoc_to_stream_prost` instead.")
+    }
+}
+
+impl StreamShare {
+    pub fn adhoc_to_stream_prost(&self, state: &mut BuildFragmentGraphState) -> ProstStreamPlan {
+        let operator_id = self.base.id.0 as u32;
+        match state.get_share_stream_node(operator_id) {
+            None => {
+                let dispatch_strategy = match &self.base.dist {
+                    Distribution::HashShard(keys) | Distribution::UpstreamHashShard(keys, _) => {
+                        DispatchStrategy {
+                            r#type: DispatcherType::Hash as i32,
+                            column_indices: keys.iter().map(|x| *x as u32).collect_vec(),
+                        }
+                    }
+                    Distribution::Single => DispatchStrategy {
+                        r#type: DispatcherType::Simple as i32,
+                        column_indices: vec![],
+                    },
+                    Distribution::Broadcast => DispatchStrategy {
+                        r#type: DispatcherType::Broadcast as i32,
+                        column_indices: vec![],
+                    },
+                    Distribution::SomeShard => {
```

Comment:
I guess this case is unreachable?

Reply:
Actually when the input of share operator is a source, it would be, since source operator's distribution is `Distribution::SomeShard`.

Reply:
You're right. The dispatcher type in `Exchange::to_stream` is based on the downstream distribution. 🥵 Sounds like the `Hash` here may lead to unnecessary shuffle and the best type here is `NoShuffle`. While according to the [discussions](https://www.notion.so/risingwave-labs/Multiple-Edges-in-the-Stream-Graph-f0ab595523d74823a39edaebdea2be16#8353c5ee42c746899c08e2d85fdf4f9e) for multiple edges, using `Hash` might be the only choice.